pages/design_documentation/cleaner.md

```diff
@@ -5,12 +5,138 @@ toc: false
+background task is then put to sleep for **Wake Up** time period.
+
+
 ### ACP - Aggressive Cleaning Policy
-### NOP - No Operation (effectively disables background cleaning)
+Aggressive Cleaning Policy takes a rather different approach than ALRU. It organizes
+the data into 100MiB chunks, maintaining a list of them per core device. The 
+chunks are put into "buckets" of varying dirty data percentage. There are 11 
+buckets in the ACP, holding data chunks being 0% dirty,  up to 100% dirty, 
+with a 10% granularity.
+
+
+#### Configuration
+ACP uses these config parameters to tune the behavior of the cleaner.
+
+* **Wake Up**- cleaning thread wake up time. (default: 10s)
+
+* **Flush Max Buffers** - maximum amount of dirty cache lines to flush in one cleaning 
+operation (default: 128 lines)
+
+
+#### Operation
+When the cache is started and core devices are added, ACP creates the chunks lists
+per each core. A chunk is a linear, 100MiB part of the core device address space,
+cache-line addressable. This creates an ACP-specific mapping of cache lines to 
+the chunks. These chunks are then put into a proper buckets, depending on the 
+percentage of dirty cache lines in a chunk.
+
+When a cache line is written and dirty data is created, the dirty lines counter 
+in the chunk the cache line belongs to, is incremented. Since the chunks are placed 
+in the buckets according to the dirty lines percentage in a chunk, a chunk may be
+moved to another bucket once the percentage threshold is met.
+
+After **Wake Up** time has passed, background cleaning thread is woken up and
+tries to perform cleaning. In the ACP, the cleaning logic selects the chunk to 
+clean by scanning backwards across the buckets, i.e. starting from the 100% bucket
+down to the 10% one. 
+As soon as there is a non-empty chunk list within a bucket, first chunk from that 
+list is selected. This chunk will be cleaned only if the core device it belongs 
+to is opened and the next cleaning timestamp is exceeded (a backoff time after 
+flush error has occurred - 5s).
+
+The list of up to **Flush Max Buffers** of cache lines is provided to the OCF
+cleaner. The *ocf_cleaner_fire()* is called and the OCF cleaner issues an 
+asynchrounous flushing request, eventually writing the dirty data to the core 
+device.
+
+After a successful flush, the cleaning thread will be woken up after **Wake Up**
+time and the selection of a chunk to clean will commence again. If the chunk 
+that was previously cleaned still has some dirty data, it will be selected again, 
+the cleaning constraints will be evaluated again and the next portion of up to
+**Flush Max Buffers** of cache lines of this chunk will be flushed.
+
 
-## Other uses
+### NOP - No Operation
+No operation is performed by the background cleaner task, no logic is called to 
+select and/or flush the data. This effectively disables the periodic background 
+cleaning. Other cleaning cases remain unaffected.
```

Comment:
So basically `ocf_cleaner_run()` will not perform any cleaning, and the callback will be always called with an `interval` set to `OCF_CLEANER_DISABLE` (which happens to be `UINT_MAX`, so someone who doesn't handle it explicitly, may experience some veeery long waiting thread).